Simplified/Reader2/NYPLReaderBookmarksBusinessLogic.swift

```diff
@@ -200,25 +227,119 @@ class NYPLReaderBookmarksBusinessLogic: NSObject, NYPLReadiumViewSyncManagerDele
+          {
+            self.bookmarks = self.bookRegistry.readiumBookmarks(forIdentifier: self.book.identifier)
+            completion(true, self.bookmarks)
+          }
+        }
+      }
+      
+    })
+  }
+    
+  private func updateLocalBookmarks(serverBookmarks: [NYPLReadiumBookmark],
+                                     localBookmarks: [NYPLReadiumBookmark],
+                                     bookmarksFailedToUpload: [NYPLReadiumBookmark],
+                                     completion: @escaping () -> ())
+  {
+    // Bookmarks that are present on the client, and have a corresponding version on the server
+    // with matching annotation ID's should be kept on the client.
+    var localBookmarksToKeep = [NYPLReadiumBookmark]()
+    // Bookmarks that are present on the client, have been uploaded before,
+    // but are no longer on the server, should be deleted on the client.
+    var localBookmarksToDelete = [NYPLReadiumBookmark]()
+    // Bookmarks that are present on the server, but not the client, should be added to this
+    // client as long as they were not created on this device originally.
+    var serverBookmarksToKeep = serverBookmarks
+    // Bookmarks present on the server, that were originally created on this device,
+    // and are no longer present on the client, should be deleted on the server.
+    var serverBookmarksToDelete = [NYPLReadiumBookmark]()
+    
+    for serverBookmark in serverBookmarks {
+      let matchingBookmarks = localBookmarks.filter { $0.annotationId == serverBookmark.annotationId }
+        
+      localBookmarksToKeep.append(contentsOf: matchingBookmarks)
+        
+      if let deviceID = serverBookmark.device,
+        deviceID == NYPLUserAccount.sharedAccount().deviceID
+        && matchingBookmarks.count == 0
+      {
+        serverBookmarksToDelete.append(serverBookmark)
+        if let indexToRemove = serverBookmarksToKeep.index(of: serverBookmark) {
+          serverBookmarksToKeep.remove(at: indexToRemove)
+        }
+      }
+    }
+    
+    for localBookmark in localBookmarks {
+      if !localBookmarksToKeep.contains(localBookmark) {
+        bookRegistry.delete(localBookmark, forIdentifier: self.book.identifier)
+        localBookmarksToDelete.append(localBookmark)
+      }
+    }
+    
+    var bookmarksToAdd = serverBookmarks
+    bookmarksToAdd.append(contentsOf: bookmarksFailedToUpload)
+        
+    for serverBookmark in serverBookmarksToKeep {
+      for localBookmark in localBookmarksToKeep {
+        if serverBookmark.isEqual(localBookmark), let index = bookmarksToAdd.index(of: localBookmark) {
+          bookmarksToAdd.remove(at: index)
+        }
+      }
+    }
+        
+    for bookmark in bookmarksToAdd {
+      bookRegistry.add(bookmark, forIdentifier: self.book.identifier)
+    }
+    
+    if serverBookmarksToDelete.count > 0 {
+      NYPLAnnotations.deleteBookmarks(serverBookmarksToDelete)
```

Comment:
should this call be moved out of this function, since this function is about updating the local bookmarks?

Reply:
The only place I can think of is the completion block in `syncBookmarks`, which would make it more complicated, what do you think? I removed the if-statement though since I don't think it's necessary